packages/flutter/lib/src/services/hardware_keyboard.dart

```diff
@@ -0,0 +1,894 @@
+/// These states will have been updated with the event when used during a key
+/// event handler.
+///
+/// The singleton [HardwareKeyboard] instance is held by the [ServicesBinding]
+/// as [ServicesBinding.keyboard], and can be conveniently accessed using the
+/// [HardwareKeyboard.instance] static accessor.
+///
+/// ## Event model
+///
+/// Flutter uses a universal event model ([KeyEvent]) and key options
+/// ([LogicalKeyboardKey] and [PhysicalKeyboardKey]) regardless of the native
+/// platform, while preserving platform-specific features as much as
+/// possible.
+///
+/// [HardwareKeyboard] guarantees the key model to be "regularized": The key
+/// event stream consists of "key tap sequences", where a key tap sequence is
+/// defined as one [KeyDownEvent], zero or more [KeyRepeatEvent]s, and one
+/// [KeyUpEvent] in order, all with the same physical key and logical key.
+///
+/// When the Flutter application starts, all keys are released, and all lock
+/// modes are disabled. Upon key events, [HardwareKeyboard] will update its
+/// states, then dispatch callbacks: [KeyDownEvent]s and [KeyUpEvent]s set
+/// or reset the pressing state, while [KeyDownEvent]s also toggle lock modes.
+///
+/// Flutter will try to synchronize with the ground truth of keyboard states
+/// using synthesized events ([KeyEvent.synthesized]), subject to the
+/// availability of the platform. The desynchronization can be caused by
+/// non-empty initial state or loss of focus.  For example, if CapsLock is
+/// enabled when the application starts, then immediately before the first key
+/// event, a synthesized [KeyDownEvent] and [KeyUpEvent] of CapsLock will be
+/// dispatched.
+///
+/// The resulting event stream does not map one-to-one to native key event
+/// stream. Some native events might be skipped, while some events
+/// might be fabricated and do not correspond to native events, marked
+/// by [KeyEvent.synthesized].
+///
+/// Flutter does not distinguish between multiple keyboards. Flutter will
+/// process all events as if they come from a single keyboard, and try to
+/// resolve any conflicts and provide a regularized key event stream.
+///
+/// ## Compared to [RawKeyboard]
+///
+/// [RawKeyboard] is the legacy API, will be deprecated and removed in the
+/// future.
+///
+/// [RawKeyboard] dispatches events that contain raw key event data and converts
+/// them to [RawKeyEvent]. [RawKeyboard] provides a less unified, less regular
+/// event model than [HardwareKeyboard].
+///
+/// It is recommended to always use [HardwareKeyboard] and [KeyEvent] APIs to
+/// handle key events.
+///
+/// See also:
+///
+///  * [KeyDownEvent], [KeyRepeatEvent], and [KeyUpEvent], the classes used to
+///    describe specific key events.
+///  * [instance], the singleton instance of this class.
+///  * [RawKeyboard], the legacy API that dispatches key events containing raw
+///    system data.
+class HardwareKeyboard {
```

Comment:
What if we remove `Hardware`? I've always bothered by how long this name is. I understand that this might confuse with on-screen keyboards, but I think this is something that can be explained.

Also we might use `HardwareKeyboardState`, since this class only has state getters besides adding and removing listeners.

Reply:
I actually like the "Hardware" part: I've had several bugs filed in the past where people were confused when the `RawKeyboard` class didn't provide events from software keyboards (even though it said so in the documentation). I think including "Hardware" here helps eliminate that confusion.

It's OK, I don't think we need to change the name, it's not that likely that people will assume that they can control the hardware keyboard itself.